vault/seal/transit/transit.go

```diff
@@ -0,0 +1,233 @@
+package transit
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+	"sync/atomic"
+
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/api"
+	"github.com/hashicorp/vault/physical"
+	"github.com/hashicorp/vault/vault/seal"
+)
+
+// Seal is a seal that leverages Vault's Transit secret
+// engine
+type Seal struct {
+	logger  log.Logger
+	client  *api.Client
+	renewer *api.Renewer
+
+	mountPath string
+	keyName   string
+
+	currentKeyID *atomic.Value
+}
+
+var _ seal.Access = (*Seal)(nil)
+
+// NewSeal creates a new transit seal
+func NewSeal(logger log.Logger) *Seal {
+	s := &Seal{
+		logger:       logger.ResetNamed("seal-transit"),
+		currentKeyID: new(atomic.Value),
+	}
+	s.currentKeyID.Store("")
+	return s
+}
+
+// SetConfig processes the config info from the server config
+func (s *Seal) SetConfig(config map[string]string) (map[string]string, error) {
+	if config == nil {
+		config = map[string]string{}
+	}
+
+	switch {
+	case os.Getenv("VAULT_TRANSIT_SEAL_MOUNT_PATH") != "":
+		s.mountPath = os.Getenv("VAULT_TRANSIT_SEAL_MOUNT_PATH")
+	case config["mount_path"] != "":
+		s.mountPath = config["mount_path"]
+	default:
+		return nil, fmt.Errorf("mount_path is required")
+	}
+
+	switch {
+	case os.Getenv("VAULT_TRANSIT_SEAL_KEY_NAME") != "":
+		s.keyName = os.Getenv("VAULT_TRANSIT_SEAL_KEY_NAME")
+	case config["key_name"] != "":
+		s.keyName = config["key_name"]
+	default:
+		return nil, fmt.Errorf("key_name is required")
+	}
+
+	var disableRenewal bool
+	var disableRenewalRaw string
+	switch {
+	case os.Getenv("VAULT_TRANSIT_SEAL_DISABLE_RENEWAL") != "":
+		disableRenewalRaw = os.Getenv("VAULT_TRANSIT_SEAL_DISABLE_RENEWAL")
+	case config["disable_renewal"] != "":
+		disableRenewalRaw = config["disable_renewal"]
+	}
+	if disableRenewalRaw != "" {
+		var err error
+		disableRenewal, err = strconv.ParseBool(disableRenewalRaw)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var namespace string
+	switch {
+	case os.Getenv("VAULT_NAMESPACE") != "":
+		namespace = os.Getenv("VAULT_NAMESPACE")
+	case config["namespace"] != "":
+		namespace = config["namespace"]
+	}
+
+	apiConfig := api.DefaultConfig()
+	if config["address"] != "" {
+		apiConfig.Address = config["address"]
+	}
```

Comment:
I think we also need to allow configuration of client TLS certs, it's likely that mTLS will be required.